### ISSUE TITLE
Circular dependencies are not detected

### ISSUE THREAD

**Issue by Dragnalith** [gracicot/kangaru#77]
Circular Dependencies are not detected. My test application crash on a segfault after recursing for ever.

The best solution would be to return are error at compile time, if not possible at runtime.

This is my main concern before using kangaru in my professional projects. On complexe service graph, other developer could create circular dependency by mistake, but without any explicit error they may not notice the problem and may not understand that's the reason the application is crashing.

**Comment by gracicot** [comment]
Thanks for reporting an issue.

Indeed, especially mixed with autocall, the library won't try to detect those errors.

A recent refactor made it easier to detect errors at compile time, but recursive type traits are quite tedious to make right. I'll see what I can do for the next version (4.2).

Also, could you post a small snippet that reproduces the error? It would help me ensure your problem is fixed and tested.

**Comment by Dragnalith** [comment]
For instance that code where all the services depends on each other.

```cpp
#include <kangaru/kangaru.hpp>

namespace my_ns
{

class CompB;
class CompA;
class CompC;

class CompC
{
public: 
    CompC(CompB& b) : m_b(b) {}

    CompB& m_b;
    int foo;
};

class CompB
{
public:
    CompB(CompA& a) : m_a(a) {}

    CompA& m_a;
    int hello = 3;
};

class CompA
{
public:
    CompA(CompC& c)
    {
    }
    int hello = 2;
};

}

struct AService : kgr::single_service<my_ns::CompA, kgr::autowire> {};
struct BService : kgr::single_service<my_ns::CompB, kgr::autowire> {};
struct CService : kgr::single_service<my_ns::CompC, kgr::autowire> {};

namespace my_ns
{

auto service_map(CompA const&) -> AService;
auto service_map(CompB const&) -> BService;
auto service_map(CompC const&) -> CService;
}

int main()
{
    kgr::container container;
    my_ns::CompA& sA = container.service<AService>();
}
```

**Comment by gracicot** [comment]
Thanks, I'll take a look a it soon. Also note that circular dependencies can be solved using setter injection:

```c++
struct CompA
{
    setC(CompC&){ }
};

struct AService : kgr::single_service<my_ns::CompA>,
    kgr::autocall<
        kgr::method<decltype(&CompA::setC), &CompA::setC>
    > {};
```

**Comment by gracicot** [comment]
Hmm... I did some experiments, and it's a lot harder than expected. When using fixed dependency, you indeed get a compile time error, since I require all services definition to be complete when specifying dependencies.

This is different for autowire. With autowire we defer the resolving of dependencies until using the serivce. At that point, not only every service are complete, but the only place we know a specific dependency is in the conversion operator of [`kgr::detail::deducer`](https://github.com/gracicot/kangaru/blob/ecbc51c35c078e3f4d00411874abe7330484cec1/include/kangaru/detail/autowire_traits.hpp#L22). As far as I know, the check is not possible to do here, and the type cannot be extracted back of the conversion operator.

Did `boost::di` solved that problem? Does it throw a runtime exception, a compile time error or just a crash? If they managed to implement that check, maybe I can borrow some of thier magic :wink:

**Comment by Dragnalith** [comment]
@gracicot Yes `boost::di` solved that problem, but I do not know how. Everything is compiled time, when I have a circular dependency I got a compile error saying I cannot construct the type I am trying to construct (but it is not obvious the reason is because of circular dependency).

Everything is compiled time, so they do not even have a `type_id` class.

**Comment by gracicot** [comment]
@Dragnalith In the branch `dev-4.2.x`, this issue has been fixed. Circular dependencies are detected by now.

Also, indeed we have a `type_id` class, because boost-di and kangaru works in different way. In our case, we went this way because the container can grow dynamically and resolve more services at runtime. A zero overhead interface like boost is in our roadmap, but not a priority yet.

**Comment by gracicot** [comment]
The fix triggered a bug in GCC. I reverted the fix until a workaround has been found.

**Comment by gracicot** [comment]
I got it working with GCC, but I'm not satisfied with the fix yet. I'll continue to work on it but this issue should be closed soon.

**Comment by gracicot** [comment]
Finally fixed in `dev-4.2.x`. Thank you again for reporting the bug!
